src/internal/restart/restart.go

```diff
@@ -0,0 +1,138 @@
+// Package restart shuts down Pachyderm when someone wants it to.
+package restart
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/admindb"
+	"github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"go.uber.org/zap"
+)
+
+var startupTime = time.Now()
+
+// checkFn is a function that checks if pachd needs to restart right now, returning whether to
+// restart and the reason for restarting.
+type checkFn func(ctx context.Context) (restart bool, reason string, err error)
+
+// restartFn is a function that restarts Pachyderm for the provided reason.
+type restartFn func(ctx context.Context, reason string)
+
+// Restarter is an object that manages restarting Pachyderm.
+type Restarter struct {
+	ch      listenerChan
+	check   checkFn
+	restart restartFn
+}
+
+// New returns a restarter suitable for use in K8s.
+func New(db *pachsql.DB, listener collection.PostgresListener) (*Restarter, error) {
+	ch := make(listenerChan)
+	if err := listener.Register(ch); err != nil {
+		return nil, errors.Wrap(err, "register restart listener")
+	}
+	return &Restarter{
+		check: func(ctx context.Context) (restart bool, reason string, err error) {
+			if err := dbutil.WithTx(ctx, db, func(ctx context.Context, tx *pachsql.Tx) error {
+				var err error
+				restart, reason, err = admindb.ShouldRestart(ctx, tx, startupTime, time.Now())
+				if err != nil {
+					return errors.Wrap(err, "ShouldRestart")
+				}
+				return nil
+			}); err != nil {
+				return false, "", errors.Wrap(err, "WithTx")
+			}
+			return restart, reason, nil
+		},
+		restart: func(ctx context.Context, reason string) {
+			log.Info(ctx, "pachyderm restart requested; restarting in 10s", zap.String("reason", reason))
+			select {
+			case <-time.After(10 * time.Second):
+			case <-ctx.Done():
+			}
+			if err := os.WriteFile("/dev/termination-log", []byte("automatic restart: "+reason), 0o777); err != nil {
+				log.Debug(ctx, "problem writing restart reason to /dev/termination-log", zap.Error(err))
+			}
+			log.Exit(ctx, "restart requested; exiting now", zap.String("reason", reason))
+		},
+		ch: ch,
+	}, nil
+}
+
+// RestartWhenRequired blocks until the context expires or the restarter callback is called.
+func (r *Restarter) RestartWhenRequired(rctx context.Context) error {
+	for {
```

Comment:
The logical flow from here through `runOnce` and into `checkAndRestart` is difficult to understand, I think because both restarts and context cancellation are represented as errors, while errors themselves are logged and ignored.

What is more surprising is that — if I understand correctly — those restart errors only appear in testing, because in a real restart the process exits first.

Reply:
checkAndRestart exists because events from two channels (timer, database change listener) do the same thing, and I want to run the same code.

checkAndRestart returns an error because an error can occur.  errors in `check` are considered transient and can retry (do not care if the table is locked for migrations or something).  errors in `restart` are considered terminal for the restarter.  Yes, in production we `os.Exit` so the function doesn't return, but representing the thing that ends polling as an error makes sense to me.

Reply:
> errors in `restart` are considered terminal for the restarter.

Are they? `restart` doesn’t return an error …

> representing the thing that ends polling as an error makes sense to me

I think it makes more sense to reserve errors for errors☺

It’s not the end of the world, this is a small piece of code which appears to act correctly, it’s just a bit surprising on a couple of different levels. Not really a blocker, though — although I think the clock skew arguably is.